google-ads/build.gradle

```diff
@@ -13,42 +16,268 @@
+  // configuration by default.
+  dependencyConfiguration = 'runtimeClasspath'
+
+  // Defines variables for the expected output files.
+  def outputDirectory = tasks.downloadLicenses.outputs.files.files.find()
+  def outputFilePattern = outputDirectory.absoluteFile.path + File.separator + "dependency-license."
+  def licenses = ["html", "json", "xml"].collect({ file(outputFilePattern + it) })
+
+  // Exports the output file names for use elsewhere.
+  ext {
+    licenseFiles = licenses
+  }
+}
+
+task verifyLicenses() {
+  description = "Verifies that the licenses used by 3P dependencies are allowed"
+  dependsOn 'generateLicenses'
+
+  doLast {
+    // Reads the dependency-license.json file generated by generateLicenses.
+    def inputFile = tasks.generateLicenses.licenseFiles.find { it.path.endsWith(".json") }
+    def json = new JsonSlurper().parseText(inputFile.text)
+    // Gets all the licenses except those for ourselves (which aren't set yet).
+    def licenses = json
+        .dependencies
+        .findAll { !it.name.startsWith("com.google.api-ads") }
+        .licenses
+        .name
+        .flatten() as Set
+    // Defines licenses which are allowed. Update this list if you need to add
+    // a new license.
+    def allowedLicenses = [
+        "Apache License, Version 2.0",
+        "The Apache License, Version 2.0",
+        "The Apache Software License, Version 2.0",
+        "Apache 2.0",
+        "Apache-2.0",
+        "Apache 2",
+        "BSD",
+        "BSD 3-clause",
+        "3-Clause BSD License",
+        "The MIT License",
+        "MIT License",
+        "MIT license",
+        "BSD New license",
+        "GNU General Public License, version 2 (GPL2), with the classpath exception",
+        "CDDL + GPLv2 with classpath exception"
+    ] as Set
+    // Computes which licenses are in use but not permitted.
+    def notAllowed = Sets.difference(licenses, allowedLicenses)
+    if (!notAllowed.empty) {
+      throw new GradleException(
+          """Licenses found which were not allowed: ${notAllowed}
+- Note that this script is not very smart. It doesn't attempt to do partial
+string comparison on the license name. It is possible that the new license is
+already allowed with a different spelling. If the new license is already allowed
+just add the new spelling to google-ads/build.gradle.
+- Otherwise you need to consider if we should accept this new license. We only
+allow licenses which are listed on go/thirdpartylicenses as notice, reciprocal
+or permissive type. The only restricted license allowed is GPL v2 with the
+classpath exception. If the new licenses are permitted you can update
+google-ads/build.gradle. Otherwise talk to go/opensource or don't add it.""")
+    }
+  }
+}
+
+task copyLicenses(type: Copy) {
+  description = "Copies licenses from build directory to third_party"
+  dependsOn 'generateLicenses'
+
+  from tasks.generateLicenses.licenseFiles
+  into '../third_party'
```

Comment:
Is this automatically named or a gradle convention? Just wondering why we're using third_party instead of ThirdParty.

Reply:
It's the convention suggested by Open Source Program Office, see http://shortn/_aAotX5Gtum